File: exercise/C#/day22/EID/SerialNumber.cs
using LanguageExt;

namespace EID
{
    public readonly struct SerialNumber : IEquatable<SerialNumber>
    {
        private readonly int _value;
        public SerialNumber(int value) => _value = value;

        public static explicit operator SerialNumber(int value) =>
            Parse(value.ToString())
                .Match(s => s, err => throw new InvalidCastException(err.Reason));

        public static Either<ParsingError, SerialNumber> Parse(string potentialSerialNumber)
            => potentialSerialNumber.ToInt()
                .Filter(x => x is >= 1 and <= 999)
                .Map(number => new SerialNumber(number))
                .ToEither(new ParsingError("serial number should be between 1 and 999"));

        public override string ToString() => $"{_value:D3}";
        public bool Equals(SerialNumber other) => _value == other._value;
        public override bool Equals(object? obj) => obj is Year other && Equals(other);
        public override int GetHashCode() => _value;
        public static bool operator ==(SerialNumber left, SerialNumber right) => left.Equals(right);
        public static bool operator !=(SerialNumber left, SerialNumber right) => !(left == right);
    }
}

File: exercise/ts/day22/src/eid/serialNumber.ts
import {Either, left, right} from 'fp-ts/Either';
import {ParsingException} from "./parsingException";

export class SerialNumber {
    private constructor(private readonly value: number) {
    }

    static parse(input: string): Either<ParsingException, SerialNumber> {
        const number = parseInt(input);
        if (isNaN(number) || number < 1 || number > 999) {
            return left(new ParsingException('Serial number should be between 1 and 999'));
        }
        return right(new SerialNumber(number));
    }

    toString(): string {
        return this.value.toString().padStart(3, '0');
    }

    equals(other: SerialNumber): boolean {
        if (!(other instanceof SerialNumber)) {
            return false;
        }
        return this.value === other.value;
    }
}